### PR TITLE
Added UART assisted test RPI scripts

### PR DESCRIPTION
**Pull request by lundinc2**
Added UART assisted test RPI scripts

Description
-----------

Checklist:
----------

- [ ] I have tested my changes. No regression in existing tests.
- [ ] My code is Linted.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: libraries/abstractions/common_io/test/test_iot_uart.c [hunk 1]
```diff
@@ -399,44 +402,41 @@ TEST( TEST_IOT_UART, AFQP_IotUARTWriteAsyncReadAsyncLoopbackTest )
...
     uint8_t cpBufferLarge[ testIotUART_BUFFER_LENGTH_LARGE + 1 ] = { 0 };
-    uint8_t cpBufferRead[ testIotUART_BUFFER_LENGTH_LARGE + 1 ] = { 0 };
-    uint8_t uStringCompare = 1;
+    BaseType_t xCallbackReturn;
 
     for (i = 0; i < testIotUART_BUFFER_LENGTH_LARGE; i++)
     {
         cpBufferLarge[i] = 0xAA;
     }
+    strncpy(&cpBufferLarge[testIotUART_BUFFER_LENGTH_LARGE ], "\n", 1);
```
Comment (qiutongs) [inline libraries/abstractions/common_io/test/test_iot_uart.c]: nit: you might just want:

```
cpBufferLarge[testIotUART_BUFFER_LENGTH_LARGE ] = '\n'
```

File: libraries/abstractions/common_io/test/test_iot_uart.c [hunk 2]
```diff
@@ -537,7 +537,7 @@ TEST( TEST_IOT_UART, AFQP_IotUARTCancel )
     int32_t lIoctl, lWrite, lClose, lCancel;
     BaseType_t xCallbackReturn;
 
-    uint8_t cSmallBuf[ 2 ] = { 'H', 'I' };
+    uint8_t cSmallBuf[ 3 ] = { 'H', 'I', '\n' };
```
Comment (qiutongs) [inline libraries/abstractions/common_io/test/test_iot_uart.c]: nit, you can:

```
uint8_t cSmallBuf[] = "HI"
```

File: libraries/abstractions/common_io/test/test_iot_uart.c [hunk 3]
```diff
@@ -570,6 +570,8 @@ TEST( TEST_IOT_UART, AFQP_IotUARTCancel )
     lWrite = iot_uart_write_async( xUartHandle, cSmallBuf, uSmallBuflen );
     TEST_ASSERT_EQUAL( IOT_UART_SUCCESS, lWrite );
 
+    vTaskDelay(pdMS_TO_TICKS(1000));
```
Comment (qiutongs) [inline libraries/abstractions/common_io/test/test_iot_uart.c]: why delay 1 seconds here?

File: libraries/abstractions/common_io/test/test_scripts/uart/test_iot_uart_rp3.py [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+def write_async():
+    rcv = serialport.read_until()
+    print(rcv)
+
+
+def write_read_sync():
+    rcv = serialport.read_until()
+    rcv_str = str(repr(rcv))
+    print(rcv_str)
+    serialport.write(rcv)
```
Comment (qiutongs) [inline libraries/abstractions/common_io/test/test_scripts/uart/test_iot_uart_rp3.py]: Can you add some comments? I don't quite understand the logic here due to lack of experience with python and RPI...

File: libraries/abstractions/common_io/test/test_iot_uart.c [hunk 5]
```diff
@@ -371,23 +373,24 @@ TEST( TEST_IOT_UART, AFQP_IotUARTWriteAsyncReadAsyncLoopbackTest )
...
-
         lRead = iot_uart_read_async( xUartHandle, cpBufferRead, testIotUART_BUFFER_LENGTH );
         TEST_ASSERT_EQUAL( IOT_UART_SUCCESS, lRead );
 
         /* Delay for 1 sec. */
         vTaskDelay(testIotUART_DELAY);
 
+        lIoctl = iot_uart_ioctl( xUartHandle, eGetTxNoOfbytes, &lTransferAmount );
+        TEST_ASSERT_EQUAL( IOT_UART_SUCCESS, lIoctl );
+        TEST_ASSERT_GREATER_THAN( 0, lTransferAmount  );
```
Comment (vjjunnut) [inline libraries/abstractions/common_io/test/test_iot_uart.c]: What is the reason to move it after read?

File: libraries/abstractions/common_io/test/test_iot_uart.c [hunk 6]
```diff
@@ -399,44 +402,41 @@ TEST( TEST_IOT_UART, AFQP_IotUARTWriteAsyncReadAsyncLoopbackTest )
...
-        TEST_ASSERT_EQUAL( IOT_UART_SUCCESS, lIoctl );
-        TEST_ASSERT_GREATER_THAN( 0, lTransferAmount_1  );
-
-        /* Delay for 1 sec. */
-        vTaskDelay(testIotUART_DELAY);
+        xCallbackReturn = xSemaphoreTake( xtestIotUARTSemaphore, testIotUART_DEFAULT_SEMPAHORE_DELAY );
+        TEST_ASSERT_EQUAL( pdTRUE, xCallbackReturn );
 
-        lIoctl = iot_uart_ioctl( xUartHandle, eGetTxNoOfbytes, &lTransferAmount_2 );
+        lIoctl = iot_uart_ioctl( xUartHandle, eGetTxNoOfbytes, &lTransferAmount );
```
Comment (vjjunnut) [inline libraries/abstractions/common_io/test/test_iot_uart.c]: Same question here. Why is this moved after read is completed? At this point lTranferAmount should be the size of the buffer anyway. 
Also earlier, we were making sure this can be checked multiple times and it increases

File: libraries/abstractions/common_io/test/test_iot_uart.c (hunk 1 above)
Comment (lundinc2) [reply libraries/abstractions/common_io/test/test_iot_uart.c]: okay

File: libraries/abstractions/common_io/test/test_iot_uart.c (hunk 3 above)
Comment (lundinc2) [reply libraries/abstractions/common_io/test/test_iot_uart.c]: When I ran this case the operation was not finished when so the cancel would be successful

File: libraries/abstractions/common_io/test/test_iot_uart.c (hunk 2 above)
Comment (lundinc2) [reply libraries/abstractions/common_io/test/test_iot_uart.c]: Okay, need to keep the \n for the assisted test

File: libraries/abstractions/common_io/test/test_iot_uart.c (hunk 5 above)
Comment (lundinc2) [reply libraries/abstractions/common_io/test/test_iot_uart.c]: I will revert this

File: libraries/abstractions/common_io/test/test_scripts/uart/test_iot_uart_rp3.py (hunk 4 above)
Comment (lundinc2) [reply libraries/abstractions/common_io/test/test_scripts/uart/test_iot_uart_rp3.py]: Okay

File: libraries/abstractions/common_io/test/test_iot_uart.c (hunk 6 above)
Comment (lundinc2) [reply libraries/abstractions/common_io/test/test_iot_uart.c]: Will revert this too

File: libraries/abstractions/common_io/test/test_iot_uart.c (hunk 1 above)
Comment (brianfr-amazon) [reply libraries/abstractions/common_io/test/test_iot_uart.c]: No.  That is incorrect.  You want a NULL terminated string .

File: libraries/abstractions/common_io/test/test_iot_uart.c (hunk 3 above)
Comment (brianfr-amazon) [reply libraries/abstractions/common_io/test/test_iot_uart.c]: Instead of the delay, would it be better to utilize the callback?

Comment (lundinc2) [reply libraries/abstractions/common_io/test/test_iot_uart.c]: Agreed, I will replace this with a semaphore

File: libraries/abstractions/common_io/test/test_scripts/uart/test_iot_uart_test.py [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+import os, sys
+import re
+
+parentdir = os.path.dirname(os.getcwd())
+sys.path.insert(0, parentdir)
+from test_iot_test_template import test_template
+
+
+class uart_test(test_template):
+    """
```
Comment (cmcgcc) [inline libraries/abstractions/common_io/test/test_scripts/uart/test_iot_uart_test.py]: Satya suggested that class names usually follow UpperCaseCamelCase convention. Could you update it to TestUartAssisted? The tests I submitted will be updated in the new pr as well.

Comment (lundinc2) [reply libraries/abstractions/common_io/test/test_scripts/uart/test_iot_uart_test.py]: Yes will do this. Saw that a change was made to use Sockets instead of serial, I will need some help understanding / setting up this new flow.

File: libraries/abstractions/common_io/test/test_scripts/uart/test_iot_uart_test.py [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+        self._serial.reset_input_buffer()
+
+        cmd = "iot_tests test 1 0"
+        self._serial.write("\r\n".encode("utf-8"))
+        self._serial.write(cmd.encode("utf-8"))
+        self._serial.write("\r\n".encode("utf-8"))
+
+        dut_res = str(self._serial.read(100))
+        dut_result = dut_res_pattern.search(dut_res)
+
```
Comment (cmcgcc) [inline libraries/abstractions/common_io/test/test_scripts/uart/test_iot_uart_test.py]: You may want to add ipc to synchronize between processes on rpi and dut. The latency of ssh can vary among locations. The output file may not ready at this point. I updated my tests with socket in this pr https://github.com/aws/amazon-freertos/pull/1356/files.

File: libraries/abstractions/common_io/test/test_scripts/uart/test_iot_uart_test.py (hunk 7 above)
Comment (cmcgcc) [reply libraries/abstractions/common_io/test/test_scripts/uart/test_iot_uart_test.py]: Feel free to let me know if you have questions.

Comment (lundinc2) [reply libraries/abstractions/common_io/test/test_scripts/uart/test_iot_uart_test.py]: Great, thank you!

File: libraries/abstractions/common_io/test/test_scripts/uart/test_iot_uart_test.py (hunk 8 above)
Comment (lundinc2) [reply libraries/abstractions/common_io/test/test_scripts/uart/test_iot_uart_test.py]: Okay will take a look
